drivers/stepper/gpio_stepper/gpio_stepper_controller.c

```diff
@@ -0,0 +1,193 @@
+/*
+ * SPDX-FileCopyrightText: Copyright (c) 2024 Carl Zeiss Meditec AG
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT gpio_stepper
+
+#include "zephyr/drivers/gpio.h"
+#include <zephyr/kernel.h>
+#include "zephyr/drivers/stepper.h"
+#include <zephyr/logging/log.h>
+#include "stdlib.h"
+
+LOG_MODULE_REGISTER(gpio_stepper_motor_controller, CONFIG_STEPPER_LOG_LEVEL);
+
+struct gpio_stepper_motor_controller_data {
+	struct k_work stepper_work;
+	const struct gpio_dt_spec *gpio_in_pins;
+	int32_t min_position;
+	int32_t max_position;
+	int32_t actual_position;
+	int32_t target_position;
+	uint32_t actual_speed;
+	uint32_t delay_in_us;
+	bool is_moving;
+} stepper_data;
+
+struct gpio_stepper_motor_controller_config {
+	const uint16_t motor_steps_per_rev;
+	const double motor_gear_ratio;
+	const uint8_t number_in_pins;
+};
+
+static int32_t stepper_motor_move_with_4_gpios(const struct gpio_dt_spec *gpio_spec,
+					       int step_number)
+{
+	switch (abs(step_number) % 4) {
+	case 0:
+		LOG_DBG("Step 1010 %d %d %d %d", gpio_spec[0].pin, gpio_spec[1].pin,
+			gpio_spec[2].pin, gpio_spec[3].pin);
+		gpio_pin_configure_dt(&gpio_spec[0], GPIO_OUTPUT_HIGH);
+		gpio_pin_configure_dt(&gpio_spec[1], GPIO_OUTPUT_LOW);
+		gpio_pin_configure_dt(&gpio_spec[2], GPIO_OUTPUT_HIGH);
+		gpio_pin_configure_dt(&gpio_spec[3], GPIO_OUTPUT_LOW);
+		break;
+	case 1:
+		LOG_DBG("Step 0110");
+		gpio_pin_configure_dt(&gpio_spec[0], GPIO_OUTPUT_LOW);
+		gpio_pin_configure_dt(&gpio_spec[1], GPIO_OUTPUT_HIGH);
+		gpio_pin_configure_dt(&gpio_spec[2], GPIO_OUTPUT_HIGH);
+		gpio_pin_configure_dt(&gpio_spec[3], GPIO_OUTPUT_LOW);
+		break;
+	case 2:
+		LOG_DBG("Step 0101");
+		gpio_pin_configure_dt(&gpio_spec[0], GPIO_OUTPUT_LOW);
+		gpio_pin_configure_dt(&gpio_spec[1], GPIO_OUTPUT_HIGH);
+		gpio_pin_configure_dt(&gpio_spec[2], GPIO_OUTPUT_LOW);
+		gpio_pin_configure_dt(&gpio_spec[3], GPIO_OUTPUT_HIGH);
+		break;
+	case 3:
+		LOG_DBG("Step 1001");
+		gpio_pin_configure_dt(&gpio_spec[0], GPIO_OUTPUT_HIGH);
+		gpio_pin_configure_dt(&gpio_spec[1], GPIO_OUTPUT_LOW);
+		gpio_pin_configure_dt(&gpio_spec[2], GPIO_OUTPUT_LOW);
+		gpio_pin_configure_dt(&gpio_spec[3], GPIO_OUTPUT_HIGH);
+		break;
+	}
+	return 0;
+}
+
+void stepper_work_handler(struct k_work *work)
+{
+	struct gpio_stepper_motor_controller_data *data =
+		CONTAINER_OF(work, struct gpio_stepper_motor_controller_data, stepper_work);
+	data->is_moving = true;
+	LOG_DBG("Moving Motor from %d to %d", data->actual_position, data->target_position);
+	if (data->target_position > data->actual_position) {
+		for (int i = 0; i < data->target_position - data->actual_position; i++) {
+			stepper_motor_move_with_4_gpios(data->gpio_in_pins, i);
+			k_sleep(K_USEC(data->delay_in_us));
+		}
```

Comment:
This is not allowed :) The work queue is async, you will block the entire system workqueue until the motor has stopped moving :) Use k_work_delayable() to step a single step every time the work item handler is called